common/rocksdb_env_s3.cpp

```diff
@@ -0,0 +1,432 @@
+// posix_env_
+Status S3Env::CreateDir(const std::string& name) {
+  return posix_env_->CreateDir(local_directory_ + name);
+};
+
+// S3 has no concepts of directories, so we just have to forward the request to
+// posix_env_
+Status S3Env::CreateDirIfMissing(const std::string& name) {
+  return posix_env_->CreateDirIfMissing(local_directory_ + name);
+};
+
+// S3 has no concepts of directories, so we just have to forward the request to
+// posix_env_
+Status S3Env::DeleteDir(const std::string& name) {
+  return posix_env_->DeleteDir(local_directory_ + name);
+};
+
+Status S3Env::GetFileSize(const std::string& fname, uint64_t* size) {
+  Status st;
+  if (posix_env_->FileExists(local_directory_ + fname).ok()) {
+    st = posix_env_->GetFileSize(local_directory_ + fname, size);
+  } else {
+    st = Status::NotFound();
+    auto meta_resp = s3_util_->getObjectSizeAndModTime(s3_key_prefix_ + fname);
+    if (!meta_resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when querying object size from S3: "
+                 << meta_resp.Error();
+    }
+
+    const auto& size_itr = meta_resp.Body().find("size");
+    if (size_itr != meta_resp.Body().end()) {
+      *size = size_itr->second;
+      st = Status::OK();;
+    }
+  }
+
+  return st;
+}
+
+Status S3Env::GetFileModificationTime(const std::string& fname,
+                                      uint64_t* file_mtime) {
+  Status st;
+  if (posix_env_->FileExists(local_directory_ + fname).ok()) {
+    st = posix_env_->GetFileModificationTime(local_directory_ + fname, file_mtime);
+  } else {
+    st = Status::NotFound();
+    auto meta_resp = s3_util_->getObjectSizeAndModTime(s3_key_prefix_ + fname);
+    if (!meta_resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when querying object ModTime from S3: "
+                 << meta_resp.Error();
+    }
+
+    const auto& mtime_itr = meta_resp.Body().find("last-modified");
+    if (mtime_itr != meta_resp.Body().end()) {
+      *file_mtime = mtime_itr->second;
+      st = Status::OK();
+    }
+  }
+
+  return st;
+}
+
+// The rename is not atomic. S3 does not support renaming natively.
+// Copy file to a new object in S3 and actually only IDENTITY files could be renamed.
+Status S3Env::RenameFile(const std::string& src, const std::string& target) {
+  auto resp = s3_util_->putObject(s3_key_prefix_ + target, local_directory_ + src);
+  if (!resp.Error().empty()) {
+    LOG(ERROR) << "Error happened when uploading new file to S3: "
+               << resp.Error();
+    return Status::IOError();
+  }
+
+  // Do the rename on local filesystem too
+  return posix_env_->RenameFile(s3_key_prefix_ + src, s3_key_prefix_ + target);
+}
+
+Status S3Env::LockFile(const std::string& fname, FileLock** lock) {
+  // there isn's a very good way to atomically check and create
+  // a file via libs3
+  *lock = nullptr;
+  return Status::OK();
+}
+
+Status S3Env::UnlockFile(FileLock* lock) {
+  return Status::OK();
+}
```

Comment:
what's the semantics of these functions? can you please add some comments for it. Does rocksdb backup/restore need to use these two functions?